text/3502-cargo-script.md

```diff
@@ -0,0 +1,1202 @@
+--manifest-path foo.rs` but that was confusing as it is unclear when a long
+execution time is from a slow compile or from the program.
+
+In the future, we could try to find ways to better control cargo's output so at
+least `cargo foo.rs` is less noisy but appears responsive.
+
+See
+[t-cargo zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/Re-thinking.20cargo's.20output)
+and [rust-lang/cargo#8889](https://github.com/rust-lang/cargo/issues/8889).
+
+Note: `--quiet` is inferred when piping to a file ([rust-lang/cargo#12305](https://github.com/rust-lang/cargo/pull/12305))
+
+## Executing `<stdin>`
+
+We could extend this to allow accepting single-file packages from stdin, either
+explicitly with `-` or implicitly when `<stdin>` is not interactive.
+
+## Implicit `main` support
+
+Like with doc-comment examples, we could support an implicit `main`.
+
+Ideally, this would be supported at the language level
+- Ensure a unified experience across the playground, `rustdoc`, and `cargo`
+- `cargo` can directly run files rather than writing to intermediate files
+  - This gets brittle with top-level statements like `extern` (more historical) or bin-level attributes
+
+Behavior can be controlled through editions
+
+## `[lib]` support
+
+In an effort to allow low-overhead packages in a workspace, we may also allow `[lib]`s to be defined.
+
+A single-file package may only be a `[bin]` or a `[lib]` and not both.
+
+We would support depending on these, publishing them, etc.
+
+We could add support for this in the future by
+- Using `syn` to check if a top-level `main` function exists (this is mutually exclusive with implicit `main`)
+- Check the manifest for an empty `[lib]` table
+
+## Workspace Support
+
+Allow scripts to be members of a workspace.
+
+The assumption is that this will be opt-in, rather than implicit, so you can
+easily drop one of these scripts anywhere without it failing because the
+workspace root and the script don't agree on workspace membership.  To do this,
+we'd expand `package.workspace` to also be a `bool` to control whether a
+workspace lookup is disallowed or whether to auto-detect the workspace
+- For `Cargo.toml`, `package.workspace = true` is the default
+- For single-file packages, `package.workspace = false` is the default
+
+When a workspace is specified
+- Use its target directory
+- Use its lock file
+- Be treated as any other workspace member for `cargo <cmd> --workspace`
+- Check what `workspace.package` fields exist and automatically apply them over default manifest fields
+- Explicitly require `workspace.dependencies` to be inherited
+  - I would be tempted to auto-inherit them but then `cargo rm`s gc will remove them because there is no way to know they are in use
+- Apply all `profile` and `patch` settings
+
+This could serve as an alternative to
+[`cargo xtask`](https://github.com/matklad/cargo-xtask) with scripts sharing
+the lockfile and `target/` directory.
```

Comment:
FWIW I think `cargo xtask.rs` could become an alternative to the current xtask pattern even without workspace support. It would allow moving `.cargo` back into `.gitignore` for some projects, which might make it worth it to ignore potential downsides... I'm not even sure what those would be, maybe the fact that `cargo xtask` can be run from arbitrary subdirectories of the workspace with the current `--package` alias setup; I don't think the same expectation applies to `cargo xtask.rs`.